### DISCUSSION TITLE
Why do we need .env file in NextJS project?

### DISCUSSION THREAD

**Discussion by sourab-updot** [vercel/next.js#58808] [Help]
### Summary

I have a decoupled solution and my frontend is on NextJS and backend is on Drupal and also using Razorpay on my nextjs project for payment gateway. My question is do I need to put all the secret/private keys used for Razorpay in .env file or can I use Drupal for setting up the keys for my frontend and then use them in Razorpay API calls?

**Comment by nicitaacom** [comment]
You need .env in your next js to get acces to some data from DB for example
Next js is a FullStack framework - it means you can use only next js without Drupal
If you want to create frontend on Next js and backend on Drupal in this case you need to set up your .env so - you have access to some data that user without this .env valus should not have
For example you have your API endpoint to get data from 'users' table - obviuosly you need to .env value to get acces to data in DB (if you use SQL in this case you need RLS as well)

So you need .env to get acces to data from DB that common user should not have access to (e.g emails phones passwords or any sencitive data)

Also I want to add that if you have API endpoint like `api/yourendpoint` you may don't set up API folder in your next js project and do it with Drupal so you don't need .env in your next js project (only in your API logic where you get your data from)

**Reply by sourab-updot** [reply]
@nicitaacom  So is this a standard practice to set all the private/secret keys in .env file in all JavaScript based framework?

**Reply by nicitaacom** [reply]
Yes usually developers put sensitive data like ANON_KEY SERVICE_ROLE_KEY in .env or .env.local

.env - default .env file
.env.local - env file ignored by git by default
(if you use prisma you need to use .env and put `.env` line in gitignore) 

In next you may notice that some keys in .env or .env.local have NEXT_PUBLIC_
it means that this keys may be exposed to client and nothing wrong with that (it just public key everyone may get access to in dev tools - but common user don't use dev tools so you may don't care)
